content/doc/developer/security/csp.adoc

```diff
@@ -0,0 +1,110 @@
+
+Using Content-Security-Policy (CSP), injection attacks like cross-site scripting can be prevented.
+
+// TODO Keep up to date with versions
+Unfortunately, as of Jenkins 2.360, the Jenkins classic UI is not yet compatible with the CSP directives that would allow preventing such injection attacks.
+
+This guide documents how to write UI code in a manner that is expected to be compatible with future use of CSP directives on the Jenkins UI.
+
+
+
+== Guidelines
+
+
+=== Inline JavaScript blocks
+
+Do not use inline JavaScript (JS) on the Jenkins UI, i.e., JS embedded in HTML output.
+
+This is typically done with `<script>` tags, like so:
+
+[source, html]
+<script type="text/javascript">
+alert("Hello, world!");
+</script>
+
+The guidelines in link:/doc/developer/security/xss-prevention/#passing-values-to-javascript[the documentation on XSS prevention] can be useful to pass arguments to JavaScript, or otherwise control its behavior dynamically.
+
+You can generally use https://github.com/jenkinsci/stapler/blob/master/docs/jelly-taglib-ref.adoc#adjunct[Stapler adjuncts] to load files related to UI views and ensure they are loaded only once.
+
+Exmples of this are: https://github.com/jenkinsci/jenkins/pull/6849[jenkinsci/jenkins#6849]
+
+
+=== Inline event handlers
+
+Event handlers like `onclick` or `onblur` would be defined in separate files.
+
+For this to work, the element that would have the inline event handler attribute(s) would need a class or ID by which it can be looked up from JS.
+
+Depending on how that element is added to the UI, you'd use on the the following methods to add event handlers:
+
+You can use `document.addEventListener('DOMContentLoaded', …)` when it's one or more elements that are present on the page from the moment it is loaded.
+Look up the elements by their ID or class or similar characteristics, then call `#addEventListener` on them.
+Be mindful of Jenkins's extensibility, so consider including plugin names in element class names or IDs to prevent unintentional conflicts with other plugins.
+
+Use `Behaviour#specify` to add event handlers to elements that may be dynamically added to the page, for example as part of AJAX responses.
+One common instance of this is in configuration forms: `renderOnDemand` is used by common form elements like `hetero-list` to load parts of the page only as the form is being changed.
+The code that adds content from AJAX responses dynamically to the page needs to call `Behaviour#applySubtree` on the newly added content.
+
+Examples of this are: https://github.com/jenkinsci/jenkins/pull/5514[jenkinsci/jenkins#5514]
+
+=== Legacy JavaScript `checkUrl` validation
+
+Form validation using manually specified `checkUrl` parameters currently supports inline JS for its "legacy" mode.
```

Comment:
I would make this prescriptive to match the first and last section in "Guidelines":

```suggestion
Do not use "legacy" mode form validation, which supports inline JS with manually specified `checkUrl` parameters.
```